pkg/controllers/serviceexport_controller.go

```diff
@@ -57,119 +65,148 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+	if changes.IsDeleted() {
+		if err := r.CloudMap.DeleteEndpoints(ctx, service.Namespace, service.Name, changes.Delete); err != nil {
+			r.Log.Error(err, "error when deleting endpoints from Cloud Map",
+				"namespace", cmService.Namespace, "name", cmService.Name)
 			return ctrl.Result{}, err
 		}
 	}
 
-	if !createRequired && !updateRequired && !deleteRequired {
-		log.Info("no changes to export", "namespace", svc.Namespace, "name", svc.Name)
+	if changes.IsNone() {
+		r.Log.Info("no changes to export", "namespace", service.Namespace, "name", service.Name)
 	}
 
 	return ctrl.Result{}, nil
 }
 
-func (r *ServiceExportReconciler) handleDelete(ctx context.Context, log logr.Logger, svcExport *v1alpha1.ServiceExport) (ctrl.Result, error) {
-	if controllerutil.ContainsFinalizer(svcExport, serviceExportFinalizer) {
+func (r *ServiceExportReconciler) createOrGetCloudMapService(ctx context.Context, service *v1.Service) (*model.Service, error) {
+	cmService, err := r.CloudMap.GetService(ctx, service.Namespace, service.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	if cmService == nil {
+		if err := r.CloudMap.CreateService(ctx, service.Namespace, service.Name); err != nil {
+			r.Log.Error(err, "error with creating a new service in Cloud Map",
+				"namespace", service.Namespace, "name", service.Name)
+			return nil, err
+		}
+		if cmService, err = r.CloudMap.GetService(ctx, service.Namespace, service.Name); err != nil {
+			return nil, err
+		}
+	}
+
+	return cmService, nil
+}
+
+func (r *ServiceExportReconciler) handleDelete(ctx context.Context, serviceExport *v1alpha1.ServiceExport) (ctrl.Result, error) {
+	if controllerutil.ContainsFinalizer(serviceExport, ServiceExportFinalizer) {
 
-		log.Info("removing Cloud Map service", "namespace", svcExport.Namespace, "name", svcExport.Name)
+		r.Log.Info("removing service export", "namespace", serviceExport.Namespace, "name", serviceExport.Name)
 
-		srv, err := r.Cloudmap.GetService(ctx, svcExport.Namespace, svcExport.Name)
+		cmService, err := r.CloudMap.GetService(ctx, serviceExport.Namespace, serviceExport.Name)
 		if err != nil {
-			log.Error(err, "error when fetching service from Cloud Map API",
-				"namespace", svcExport.Namespace, "name", svcExport.Name)
+			r.Log.Error(err, "error while fetching service from Cloud Map API",
+				"namespace", serviceExport.Namespace, "name", serviceExport.Name)
 			return ctrl.Result{}, err
 		}
-		if srv != nil {
-			if err := r.Cloudmap.DeleteEndpoints(ctx, srv.Namespace, srv.Name, srv.Endpoints); err != nil {
-				log.Error(err, "error when deleting endpoints from Cloud Map",
-					"namespace", srv.Namespace, "name", srv.Name)
+		if cmService != nil {
+			if err := r.CloudMap.DeleteEndpoints(ctx, cmService.Namespace, cmService.Name, cmService.Endpoints); err != nil {
+				r.Log.Error(err, "error while deleting endpoints from Cloud Map",
+					"namespace", cmService.Namespace, "name", cmService.Name)
 				return ctrl.Result{}, err
```

Comment:
I guess in the future we'll return something other than empty struct.

Reply:
Right now our strategy seems to be just log error, and move on. We have to probably figure out fault-tolerance strategy before GA.